src/search.rs

```diff
@@ -0,0 +1,146 @@
+use crate::search_error::SearchError;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+pub fn search(directory: &Path) -> Result<PathBuf, SearchError> {
+  let mut files = Vec::new();
+  let dir = fs::read_dir(directory).map_err(|io_error| SearchError::Io {
+    io_error,
+    directory: directory.to_owned(),
+  })?;
+  for entry in dir {
+    let entry = entry.map_err(|io_error| SearchError::Io {
+      io_error,
+      directory: directory.to_owned(),
+    })?;
+    if let Some(name) = entry.file_name().to_str() {
+      use caseless::default_caseless_match_str;
+      if default_caseless_match_str(name, "justfile") {
+        files.push(entry.path());
+      }
+    }
+  }
+  if files.len() == 1 {
+    Ok(files.pop().unwrap())
+  } else if files.len() > 1 {
+    Err(SearchError::MultipleCandidates { candidates: files })
+  } else if let Some(parent_dir) = directory.parent() {
+    search(parent_dir)
+  } else {
+    Err(SearchError::NotFound)
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use crate::search::search;
+  use crate::search_error::SearchError;
+  use std::fs;
+  use tempdir::TempDir;
+
+  #[test]
+  fn not_found() {
+    let tmp = TempDir::new("just-test-justfile-search")
+      .expect("test justfile search: failed to create temporary directory");
+    let path = tmp.path().to_path_buf();
+    match search(path.as_path()) {
+      Err(SearchError::NotFound) => {
+        assert!(true);
+      }
+      _ => panic!("No justfile found error was expected"),
+    }
+  }
+
+  #[test]
+  #[cfg(unix)]
```

Comment:
Instead of doing `#[cfg(unix)]` let's come up with a way to determine if we're running on a case insensitive filesystem, and thus the test should be skipped.

We can do:

1. Create a file called `justfile`
2. Try to open it using the name `JUSTFILE`
3. If 2. succeeds, we're on a case insensitive filesystem and the test should be skipped.